### PR TITLE
[full-ci] Enhancement: Add whitespace context-menu

### PR DESCRIPTION
**Pull request by lookacat**
## Description
We've added a generic context-menu for right clicking on whitespace.
See https://github.com/owncloud/web/issues/5861

## Related Issue

- Fixes https://github.com/owncloud/web/issues/5861

## Screenshots (if appropriate):
Screenshot (32)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [x] https://github.com/owncloud/web/pull/8938

### CODE REVIEW

**Comment by lookacat** [comment]
Refactoring of the CreateAndUpload component into composables needed before this PR
https://github.com/owncloud/web/pull/8938

File: packages/web-app-files/src/views/spaces/GenericSpace.vue [hunk 1]
```diff
@@ -393,13 +397,36 @@ export default defineComponent({
...
+        })
     })
 
     onBeforeUnmount(() => {
       visibilityObserver.disconnect()
       eventBus.unsubscribe('app.files.list.load', loadResourcesEventToken)
     })
 
+    const genericContextMenu = ref(null)
+    const showContextMenu = (event) => {
```
Comment (lookacat) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:422]: someone mentioned it would be better to do all this logic in one component which i am not a fan of since this logic toggles the whole component

**Comment by tbsbdr** [comment]
cool one 👍
A remark on Icon+wording: I'd suggest du directly show the folder icon (instead of the "+") and just call it "New Folder" (`create new` is semantically redundant)

image

**Review by kulmann — CHANGES_REQUESTED** [review]
Didn't look into the code, yet, but tried it out. Really awesome stuff! Two things that I'd like to have different:
- click on whitespace should clear the selection
- click on "details" in the whitespace context menu doesn't show the details of the current folder

Could you work on that as well?

**Comment by lookacat** [comment]
@kulmann to the first point yes possible, the second point: Did you try it in the space root? in subfolders it should work, my assumption was there is no detail on the e.g. personal space root? Or did i miss something?

**Comment by lookacat** [comment]
@kulmann alright shouldl be doable, the first point was right, i just fixed it ^^

File: packages/web-app-files/src/views/spaces/GenericSpace.vue [hunk 2]
```diff
@@ -393,13 +397,36 @@ export default defineComponent({
           performLoaderTask(true, path, fileId)
         }
       )
+      const filesViewWraapper = document.getElementsByClassName('files-view-wrapper')[0]
```
Comment (JammingBen) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:400]:
```suggestion
      const filesViewWrapper = document.getElementsByClassName('files-view-wrapper')[0]
```

File: packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue
Comment (kulmann) [inline packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue:1]: Why did you decide to not use the `ContextActionMenu.vue` component? 🤔

File: packages/web-app-files/src/views/spaces/GenericSpace.vue (hunk 2 above)
Comment (kulmann) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:400]:
```suggestion
      const filesViewWrapper = document.getElementsByClassName('files-view-wrapper')[0]
```

File: packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue
Comment (kulmann) [inline packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue:1]: Could you please use a component name which makes clear that it's hooked to whitespace? Don't have a good name in mind, yet though...

File: packages/web-app-files/src/views/spaces/GenericSpace.vue [hunk 3]
```diff
@@ -393,13 +397,36 @@ export default defineComponent({
...
+        showContextMenu(newEvent)
+      })
     })
 
     onBeforeUnmount(() => {
       visibilityObserver.disconnect()
       eventBus.unsubscribe('app.files.list.load', loadResourcesEventToken)
     })
 
+    const genericContextMenu = ref(null)
```
Comment (kulmann) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:420]: `genericContextMenu` is not a good name IMO. `whitespaceContextMenu` maybe?

File: packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue
Comment (lookacat) [reply packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue:1]: but will try to use it, shouldn't be much difference between those two besides the sections

Comment (lookacat) [reply packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue:1]: arg, my assumption was wrong, can't get the ContextActionMenu to work at this point because _tippy is always undefined, would rename it to WhitespaceContextMenu for now :/

Comment (lookacat) [reply packages/web-app-files/src/components/Spaces/SpaceGenericContextMenu.vue:1]: don't see how to quickly fix that, also there are some differences between those two component

File: packages/web-app-files/src/composables/actions/files/useFileActionsCreateNewFolder.ts [hunk 4]
```diff
@@ -115,12 +115,19 @@ export const useFileActionsCreateNewFolder = ({
...
         },
-        isEnabled: ({ resources }) => {
+        isEnabled: ({}) => {
+          if (
+            isProjectSpaceResource(space) &&
+            !space.isEditor(store.getters.user) &&
+            !space.isManager(store.getters.user)
+          ) {
+            return false
+          }
```
Comment (JammingBen) [inline packages/web-app-files/src/composables/actions/files/useFileActionsCreateNewFolder.ts:130]: This works for spaces, however, not for shares. You can simply use `space.canUpload()` I believe to cover both cases.

File: packages/web-app-files/src/views/spaces/GenericSpace.vue (hunk 2 above)
Comment (JammingBen) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:400]: There is still a typo 😄

File: packages/web-app-files/src/composables/actions/files/useFileActionsCreateNewFolder.ts (hunk 4 above)
Comment (lookacat) [reply packages/web-app-files/src/composables/actions/files/useFileActionsCreateNewFolder.ts:130]: oh thanks

File: packages/web-app-files/src/views/spaces/GenericSpace.vue (hunk 2 above)
Comment (lookacat) [reply packages/web-app-files/src/views/spaces/GenericSpace.vue:400]: oops

**Review by JammingBen — APPROVED** [review]

File: packages/web-app-files/src/composables/actions/files/useFileActionsCreateNewFolder.ts [hunk 5]
```diff
@@ -115,13 +115,13 @@ export const useFileActionsCreateNewFolder = ({
...
+        icon: 'folder',
         handler,
         label: () => {
-          return $gettext('Create new Folder')
+          return $gettext('New Folder')
         },
-        isEnabled: ({ resources }) => {
-          return true
+        isEnabled: () => {
+          return unref(currentFolder)?.canUpload({ user: store.getters.user })
```
Comment (kulmann) [inline packages/web-app-files/src/composables/actions/files/useFileActionsCreateNewFolder.ts:124]: this should use `canCreate` instead of  `canUpload`. Weirdly both `canUpload` and  `canCreate` are implemented in the same way (they check if the dav permission for folder creation is present), so you could argument that it doesn't make a difference in functionality... but semantically the creation check is the right one for folder creation. 

Side note: For the other composable (creating files) I prefer the `canUpload` check (you already have it) because if one can upload a file that's the same as creating a file. So that one should stay like it is. 👍

File: packages/web-app-files/src/views/spaces/GenericSpace.vue [hunk 6]
```diff
@@ -393,13 +397,36 @@ export default defineComponent({
           performLoaderTask(true, path, fileId)
         }
       )
+      const filesViewWrapper = document.getElementsByClassName('files-view-wrapper')[0]
+      filesViewWrapper?.addEventListener('contextmenu', (event) => {
+        const { target } = event
+        if (
+          (target as HTMLElement).closest('tbody') ||
```
Comment (kulmann) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:404]: I would prefer if you would a) define a new dummy class, b) apply that to both the tbody and all the tile list items and c) use that new dummy class as selector here. That way you don't get a dependency on specific file view modes. If we ever implement a third file view mode it would be sufficient to give the correct item your dummy class to make the whitespace context menu work.

**Review by kulmann — CHANGES_REQUESTED** [review]
Found two more things. Also, what about unit tests, at least for opening the whitespace context menu (not for the actions inside it)?

**Comment by lookacat** [comment]
@kulmann done

**Review by kulmann — APPROVED** [review]
Awesome stuff! Love the UX of this 😍
